README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# jmmm

<!-- badges: start -->
<!-- badges: end -->

The goal of jmmm is to fit multivariate mixed models and apply a
discriminant analysis frame work to predict events from the longitudinal
evolution of biomarker data.

The package was adapted from the work by Fieuws *et al.* [Fieuws
Biostatistics
2008](https://doi-org.ru.idm.oclc.org/10.1093/biostatistics/kxm041) In
summary, we fit a high-dimensional joint model by first fitting all
possible pairwise combinations of univariate mixed models within a
subset of patients with and without graft failure. Note that the subset
of subjects without the event of interest should complete the entire
follow-up, and thus no longer be at risk of the event. Subjects who did
not have the event of interest, but who have not yet accrued full
follow-up, should be excluded from model training. The pairwise mixed
models were fitted using the mixed\_model() function from the
[GLMMAdaptive package (version 0.8) developed by
Rizopoulos](https://drizopoulos.github.io/GLMMadaptive/index.html) We
join the pairwise models into a multivariate mixed model by averaging
the parameter estimates. Finally, we join the multivariate mixed models
to the event of interest using a discriminant analysis approach.
Breaking up the large multivariate model into pairwise combinations of
univariate models reduces computational burden. However, as a result of
the many possible pairwise combinations of biomarker evolutions, some
parameters for each longitudinal marker were estimated multiple times.
Therefore, estimates from the pairwise models are averaged to obtain
single estimates for all parameters in the multivariate mixed model. In
addition, subject level contributions to the first and second
derivatives for the pairwise models are averaged and used to generate a
sandwich estimator of the standard errors for the parameters. The
multivatiate mixed models for the event and survivors are used to
construct likelihood profiles for marker evolutions. In turn, the
profiles are used to generate predictions by applying Bayes’ rule.

The work was supported by a grant from the [Dutch Kidney
Foundation](https://nierstichting.nl/) - Grant 19OK003.

## Installation

You can install the development version of `jmmm` from github using the
devtools package:

`devtools::install_github("JanvandenBrand/jmmm")`

## Example

A basic example with simulated data is included below.

``` r
library(jmmm)
# The next libraries are to run the example
library(MASS)
library(gridExtra)
library(tidyverse)
#> -- Attaching packages --------------------------------------- tidyverse 1.3.1 --
#> v ggplot2 3.3.5     v purrr   0.3.4
#> v tibble  3.1.5     v dplyr   1.0.7
#> v tidyr   1.1.4     v stringr 1.4.0
#> v readr   2.0.2     v forcats 0.5.1
#> -- Conflicts ------------------------------------------ tidyverse_conflicts() --
#> x dplyr::combine() masks gridExtra::combine()
#> x dplyr::filter()  masks stats::filter()
#> x dplyr::lag()     masks stats::lag()
#> x dplyr::select()  masks MASS::select()
```

### Simulate data

First, generate some play data.

``` r
set.seed(1234)
n <- 100 # number of subjects
K <- 8 # number of measurements per subject
t_max <- 5 # maximum follow-up time

# we generate a data frame with the following design: 
# everyone has a baseline measurment, and then measurements at random follow-up times
df <- data.frame(id = as.character(rep(seq_len(n), each = K)),
                 time = c(replicate(n, c(0, sort(runif(K - 1, 0, t_max))))),
                 sex = rep(gl(2, n/2, labels = c("male", "female")), each = K))
# design matrices for the fixed effects
X1 <- model.matrix(~ sex + time, data = df)
X2 <- model.matrix(~ time, data = df)
X3 <- model.matrix(~ sex + time, data = df)
X4 <- model.matrix(~ time, data = df)
# Random intercepts model
Z1 <- 
  Z2 <- 
  Z3 <- 
  Z4 <- model.matrix(~ 1, data = df)
# fixed effects coefficients 
betas1 <- c(-0.67, -0.25, 0.24) 
betas2 <- c(-1.15, 0.3)  
betas3 <- c(1.34, -0.13, -0.04)
betas4 <- c(0.25, -0.17)
# random effects matrix
D <- c(0.8, 0.7, 0.6, 0.5)
resid <- c(0.5, 0.87) # standard deviation for error terms for y1 and y2 

# we simulate random effects
b <- cbind(rnorm(n, sd = sqrt(D[1])), rnorm(n, sd = sqrt(D[2])), rnorm(n, sd = sqrt(D[3])), rnorm(n, D[4]))
eta_y1 <- as.vector(X1 %*% betas1 + rowSums(Z1 * b[as.integer(df$id), 1, drop = FALSE]))
eta_y2 <- as.vector(X2 %*% betas2 + rowSums(Z2 * b[as.integer(df$id), 2, drop = FALSE])) 
eta_y3 <- as.vector(X3 %*% betas3 + rowSums(Z3 * b[as.integer(df$id), 3, drop = FALSE])) 
eta_y4 <- as.vector(X4 %*% betas4 + rowSums(Z4 * b[as.integer(df$id), 4, drop = FALSE])) 

# simulate normal longitudinal data
df$y1 <- rnorm(n * K, mean = eta_y1, sd = resid[1])
df$y2 <- rnorm(n * K, mean = eta_y2, sd = resid[2])
# we set binary outcome from the logistic regression
df$y3 <- as.numeric(as.logical(rbinom(n * K, size = 1, prob = plogis(eta_y3))))
df$y4 <- as.numeric(as.logical(rbinom(n * K, size = 1, prob = plogis(eta_y4))))
# Set failures
df$failure <- rnorm(n * K, mean = -0.5*eta_y1+0.8*eta_y2-0.33*eta_y3+0.15*eta_y4, sd = 1.5*mean(resid)) 
df$failure <- exp(df$failure) / (1 + exp(df$failure))
df$failure <- ifelse(df$failure >= 0.6 & df$time >0, 1, 0)
# Set failure times
subjects <- split(df, df$id)
for (i in seq_along(subjects)) {
  if ( dplyr::summarise(subjects[[i]], max(failure)) == 0 ) {
    subjects[[i]]$time_failure <- unlist(
      dplyr::summarise(subjects[[i]], max(time))
      )["max(time)"]
  } else {
    subjects[[i]]$time_failure <- unlist(
      dplyr::summarise(subjects[[i]] %>% 
                         group_by(failure) %>% 
                         dplyr::filter(failure == 1), min(time))
      )["min(time)"]
  }
  subjects[[i]]$failure <- ifelse(max(subjects[[i]]$failure) == 1, 1, 0)
  subjects[[i]] <- subjects[[i]] %>% dplyr::filter(time <= time_failure)
}
df <- do.call(rbind, subjects)
df <- df %>% mutate(sex = as.numeric(sex))
#scale the parameters
df <- df %>% mutate(y1 = scale(y1),
                    y2 = scale(y2))
```

### Initialization steps

To be able to fit the multivariate mixed model,
GLMMapative::mixed\_model() is used. However, this only accepts a single
outcome vector, not a matrix of outcomes. Therefore, we split the
outcome matrix into pairwise combinations of the longitudinal data.

``` r
pairs <- make_pairs(outcomes = c("y1", "y2", "y3", "y4"))
```

We use a support function to check for the data types of the outcomes.
The result will be used to aid in the call to fit the multivariate mixed
model later on.

``` r
model_info <- test_input_datatypes(data = df, pairs = pairs)
```

Next, we unroll the outcomes matrices of the pairwise combinations of
markers into a single outcome vector.

``` r
df_fail <- df %>% dplyr::filter(failure == 1)
df_fail_stacked <- stack_data(data = df_fail, id = "id", pairs = pairs, covars = c("time", "sex", "time_failure"))
df_nofail <- df %>% dplyr::filter(failure == 0)
df_nofail_stacked <- stack_data(data = df_nofail, id = "id", pairs = pairs, covars = c("time", "sex"))
```

Another support function can be used to construct the appropriate model
formulas.

``` r
fixed_nofail <- make_fixed_formula(covars = c("time", "sex"))
fixed_fail <- make_fixed_formula(covars = c("time", "sex", "time_failure"))
random <- make_random_formula(id = "id") 
```

### Fitting the Multivatiate Mixed Models

The actual model fitting is largely done by
GLMMadaptive::mixed\_model(). The mmm\_model() function wraps around it.
First we fit the model for the survivors. Note that the fitting
algorithm uses `future.apply` to paralellize the process in a flexible
manner.

``` r
future::plan("multisession", workers = 4)
```

The `progressr` package has been implemented to provide status updates.
You can use the following to show interactive progress updates.

    progressr::handlers(global=TRUE)
    progressr::handlers(list(
      progressr::handler_progress(
        format=":current/:total (:message) [:bar]",
        width=100,
        complete="+"
      )))

``` r
model_nofail <- mmm_model(fixed = fixed_nofail,
                          random = random,
                          id = "id",
                          data = df_nofail,
                          stacked_data = df_nofail_stacked,
                          pairs = pairs,
                          model_families = model_info,
                          iter_EM = 100,
                          iter_qN_outer = 30,
                          nAGQ = 11)
#> Fitting pairwise models:
#> Retrieving derivatives
#> Compiling model output.
```

``` r
model_fail <- mmm_model(fixed = fixed_fail,
                        random = random,
                        id = "id",
                        data = df_fail,
                        stacked_data = df_fail_stacked,
                        pairs = pairs,
                        model_families = model_info,
                        iter_EM = 100,
                        iter_qN_outer = 30,
                        nAGQ = 11)
#> Fitting pairwise models:
#> Retrieving derivatives
#> Compiling model output.
#> The Variance-Covariance matrix was not a positive definite.
#>             Projecting to the closest positive definite.
#>             Note that your results may be unexpected.
```

### Generate predictions

After the multivariate mixed model has been fitted the
mmm\_predictions() function is used to generate predicted probabilities
of the event of interest at various prediction landmarks. First we need
to generate priors though. In addition, the helper function
`get_outcome_type` is used to be able to generate appropriate
predictions and plots.

``` r
outcomes <- c("y1", "y2", "y3", "y4")
mu <- setNames(rep(0, length(outcomes)), outcomes)
re_samples_nofail <- MASS::mvrnorm(1e3,mu = mu, Sigma = model_nofail$vcov)
re_samples_fail <- MASS::mvrnorm(1e3,mu = mu, Sigma = model_fail$vcov)
prior <- get_priors(data = df,
                    time_failure = "time_failure",
                    failure = "failure",
                    horizon = 5,
                    interval = 1/4)
outcome_types <- get_outcome_type(data = df, 
                                  outcomes)
```

Generate the predictions. Note that the formulas for the fixed and
random effects need to include the longitudinal markers as well as
baseline covariates. In order to get predictions at various times, we
simply apply the function across several landmarks. The
`mmm_predictions` function support parallel processing through the
future.apply package.

``` r
predictions <- mmm_predictions(data=df,
                               outcomes=outcome_types,
                               fixed_formula_nofail="~ y1 + y2 + y3 + y4 + time + sex",
                               random_formula_nofail="~ 1| id",
                               random_effects_nofail=re_samples_nofail,
                               parameters_nofail=model_nofail$estimates,
                               fixed_formula_fail="~ y1 + y2 + y3 + y4 + time + sex + time_failure",
                               random_formula_fail="~ 1 | id",
                               random_effects_fail=re_samples_fail,
                               parameters_fail=model_fail$estimates,
                               time="time",
                               failure="failure",
                               failure_time="time_failure",
                               prior=prior,
                               id="id",
                               landmark=1,
                               horizon=5,
                               interval=1/4)
```

### Plot predictions

The following example shows the predictions for a single subject. This
helps communicate forecasts of the probability of an event conditional
on the evolution of longitudinal data. At present the function can only
be used to plot data for a single subject. However it can generate a
list of grobs for predictions generated at different landmark times.

    plot <- plot_predictions(predictions=predictions,
                     outcomes=outcomes,
                     id="id",
                     subject="1", 
                     time="time"
    )
    grid.arrange(plot)

&lt;Add later: ROC-AUC and calibration plots.&gt;

### Debugging the fitting process

When you are handling many longitudinal outcomes, a single failure can
cause the entire `mmm_model()` to fail. If this is the case a more
modular approach is desirable to help debugging. To do so, you can use
the `get_mmm_start_values()`, `get_mmm_derivatives()`, and
`average_pairwise_models()` instead of the `mmm_model()`. The
`mmm_model()` is only wrapper function. If an error is caught by
`get_mmm_start_values`, a message is returned as a place holder instead
of a fitted model object.

    start_values <- get_mmm_start_values(fixed = fixed_nofail,
                                         random = random,
                                         stacked_data = df_nofail_stacked,
                                         pairs = pairs,
                                         model_families = model_info,
                                         iter_EM = 100,
                                         iter_qN_outer = 30,nAGQ = 11)

After debugging the placeholder can be swapped out with a fitted model
as follows:

    m <- 3
    start_values[[m]] <- unlist(
      get_mmm_start_values(fixed = fixed_nofail,
                                         random = random,
                                         stacked_data = df_nofail_stacked,
                                         pairs = matrix(pairs[m,], ncol=2, nrow=1),
                                         model_families = list(families = model_info$families[m], indicators = model_info$indicators[[m]]),
                                         iter_EM = 100,
                                         iter_qN_outer = 30,
                                         nAGQ = 11),
      recursive=FALSE)

Next derivatives can be obtained:

    suppressWarnings(
    derivatives <- get_mmm_derivatives(stacked_data = df_nofail_stacked, 
                                       id = "id", 
                                       fixed = fixed_nofail, 
                                       random = random, 
                                       pairs = pairs, 
                                       model_families = model_info, 
                                       start_values = start_values, 
                                       nAGQ = 11)
    )

The final model is composed as follows:

    model_nofail <- average_pairwise_models(
      pairs = pairs,
      model_families = model_info,
      start_values = start_values,
      derivatives = derivatives,
      data = df_nofail,
      id = "id")


NEWS.md
# jmmm 0.4

## Major changes
Exported the following functions:
* `get_mmm_start_values()`
* `get_mmm_derivatives()`

Added `average_pairwise_models()`

Updated `get_start_values()` to no longer stop execution of `mmm_model()` if fitting a model fails. Instead a placeholder message is included.

Together these changes should help debug and update the multivariate mixed model if errors occur during model fitting. For an example on how to use these functions for debugging see the README file.

# jmmm 0.3

## Minor changes
* Updated handling of `progressr` to show more accurate status updates on fitting pairwise models in parallel. 
* Added error handling to retrieve the model on which fitting failed when using a parallel back end.
* Export family function

# jmmm 0.2

## Major changes
Update to the mmm_model() function. It no longer requires the user to specify the parallel back end as an argument. Instead the user should specificy their preferred back end prior to running the function. This way users have more control over parallelization in their specific setting.

## Minor changes
* Added examples to documentation.
* Added versions to package dependencies. 

tests/testthat.R
library(testthat)
library(jmmm)

test_check("jmmm")


tests/testthat/test-check_data.R
test_that("make pairs accepts only characters", {
  #
})


R/check_data.R
# Convenience functions to check and wrangle data

#' Make pairs
#'
#' Take a character with the names of the longitudinal outcome variables and make all unique pairs
#'
#' @param outcomes a character vector with the names of longitudinal outcomes
#'
#' @export
#'
#' @returns a matrix of length(outomes)*(length(outcomes)-1)/2 rows and 2 columns with all unique pairs.
#'
#' @examples
#' pairs <- make_pairs(outcomes = c("y1", "y2", "y3", "y4"))
make_pairs <- function(outcomes) {
  N <- length(outcomes)
  Npair <- N * (N-1)/2
  pairs <- matrix(data = 0, nrow = Npair, ncol = 2)
  k <- 1
  for (i in 1:(N-1)) {
    for (j in (i+1):N) {
      pairs[k,1] <- outcomes[i]
      pairs[k,2] <- outcomes[j]
      k <- k + 1
    }
  }
  pairs
}


#' Test input data types and return model families and indicator for binary outcomes
#'
#' Takes a data.frame with longitudinal data in long format and checks if the longitudinal data are binary or numeric.
#'
#' @param data A data.frame in long format
#' @param pairs A character matrix constructed by make_pairs function.
#'
#' @export
#'
#' @return A list of length nrow(pairs) that includes
#' \describe{
#'   \item{families}{character vector with the custom model family.}
#'   \item{indicator}{a character vector with the indicator for the binary outcome if family = binary.normal}
#' }
#'
#' @examples
#' \dontrun{
#' model_info <- test_input_datatypes(data = df, pairs = pairs)
#' }
test_input_datatypes <- function (data, pairs) {

  # initialize return
  .families <- vector("character", length = nrow(pairs))
  .indicators <- vector("list", length = nrow(pairs))
  # Function body
  for (i in 1:nrow(pairs)) {
    # Assert that y1 is dichotomous
    if (all(sort(unlist(unique(data[paste(pairs[i, 1])]))) == c(0, 1), na.rm = TRUE)) {
      # Assert that y2 is  dichotomous
      if (all(sort(unlist(unique(data[paste(pairs[i, 2])]))) == c(0, 1), na.rm = TRUE)) {
        .families[i] <- "binomial"
        # Assert that y2 is numeric
      } else if (class(unlist(data[paste(pairs[i, 2])])) == "numeric") {
        .families[i] <- "binary.normal"
        .indicators[[i]] <- c(1, 0)
      }
      #  Assert that y1 is numeric
    } else if (class(unlist(data[paste(pairs[i, 1])])) == "numeric") {
      # Assert that y2 is dichotomous
      if (all(sort(unlist(unique(data[paste(pairs[i, 2])]))) == c(0, 1), na.rm = TRUE)) {
        .families[i] <- "binary.normal"
        .indicators[[i]] <- c(0, 1)
        ## Assert that y2 is numeric
      } else if (class(unlist(data[paste(pairs[i, 2])])) == "numeric")  {
        .families[i] <- "normal"
      }
    } else {
      stop("The outcomes variables do not seem to be binary or continuous")
    }
  }
  list("families" = .families, "indicators" = .indicators)
}

#' Return a list with stacked data sets for each of the outcome pairs
#'
#' Takes a data.frame in with longitudinal data to be used in the multivariate mixed model function.
#' Uses the pairwise combination created with the make_pairs() function to unroll the outcome matrix into a vector.
#' The longitudinal predictor and time invariant covariates are processed accordingly.
#'
#' @param data A data.frame with the longitudinal data in long format.
#' @param pairs A character matrix with the pairs, returned from the make_pairs function
#' @param id A character value with the variable name of the subject identifier.
#' @param covars (optional) a character vector with the names of the covariate vectors.
#'
#' @import Matrix
#' @import dplyr
#'
#' @export
#'
#' @return a list of data.frames of length nrow(pairs).
#'
#' @examples
#' \dontrun{
#' df_stacked <- stack_data(
#'   data = df,
#'   id = "id",
#'   pairs = pairs,
#'   covars = c("time", "sex", "time_failure")
#' )
#' }
stack_data <- function (data, id, pairs, covars = NULL) {
  stopifnot("Argument 'id' is missing. Use it to set subject level id." = is.character(id))
  stopifnot("'id' should refer to a character vector" = is.character(unlist(data[id])))
  if (!is.null(covars)) {
    stopifnot("'covars' should be a character vector." = is.character(covars))
  }
  if (!is.null(covars)) {
    # covars should be numeric
    if (all(sapply(data %>% dplyr::select(all_of(covars)), class) == "numeric") == FALSE) {
      stop("stack_data can only handle numeric data: '",
           paste(covars[which(sapply(data %>% dplyr::select(all_of(covars)), class) != "numeric")]),
           "' is not numeric.")
    }
  }
  ids <- rep(unlist(data[id]), each = 2)
  stacked_data <- lapply(1:nrow(pairs), function(i, ...) {
    # i <- 1
    # Unroll Y matrices into vectors
    Y <- data %>%
      dplyr::select(y1 = paste(pairs[i, 1]), y2 = paste(pairs[i, 2]))
    Y <- as.vector(t(as.matrix(Y)))
    Y <- data.frame(id = ids,
                    Y = Y)
    # Create design matrices
    ## The ordering of x1 and x2 is intentional!
    X1 <- data %>%
      dplyr::select(x1 = paste(pairs[i,2]), x2 = paste(pairs[i,1]))
    X1 <- as.vector(t(as.matrix(X1)))
    ## add fixed covariates and time to the design matrices
    X2 <- data %>%
      dplyr::select(all_of(covars))
    X2 <- as.matrix(X2)
    ## add dummies for X
    intercepts <- kronecker(rep(1, times = nrow(data)), diag(1, nrow = 2))
    X2k <- kronecker(X2, diag(1, nrow = 2))
    X2k <- cbind(intercepts, X2k)
    ## join the data
    .stacked_data <- cbind(Y, X1, X2k)
    ## set variable names
    var_names <- c(id, "Y", "X", "intercept_Y1", "intercept_Y2")
    if (!is.null(covars)) {
      .covariate_names <- sapply(covars, function(j) {
        c(paste0(j,"_Y1"), paste0(j,"_Y2"))
      })
      var_names <- c(var_names, paste(.covariate_names))
    }
    names(.stacked_data) <- var_names
    .stacked_data
  })
  stacked_data
}

#' Test to compare stacked outcome data to original outcome data
#'
#' ...
#'
#' @param data data.frame with original data in long format
#' @param stacked_data the list of data.frames returned by stack_data()
#' @param pairs A character matrix with the pairs, returned from the make_pairs function
#'
#' @export
#'
#' @import dplyr
#'
#' @return Prints a list with assertions. All should return TRUE. Otherwise something went wrong.
#'
#' @examples
#' \dontrun{
#' model_info <- test_input_datatypes(data = df, pairs = pairs, stacked_data=df_stacked)
#' }
test_compare_stacked_to_original_data <- function (data, stacked_data, pairs) {
  for (i in 1:nrow(pairs)) {
    current <- unlist(data[paste(pairs[i,1])])
    target <- stacked_data[[i]] %>%
      dplyr::filter(intercept_Y1 == 1) %>%
      dplyr::select(Y)
    check <- cbind(current, target)
    print(paste('Assert: outcome vector matches the original data for', pairs[i, 1], '=',
                all.equal(check[,1], check[,2], na.rm = TRUE, check.attributes = FALSE, use.names = FALSE))
    )
    current <- unlist(data[paste(pairs[i,2])])
    target <- stacked_data[[i]] %>%
      dplyr::filter(intercept_Y1 == 1) %>%
      dplyr::select(X)
    check <- cbind(current, target)
    print(paste('Assert: outcome vector matches the original data for', pairs[i, 2], '=',
                all.equal(check[,1], check[,2], na.rm = TRUE, check.attributes = FALSE, use.names = FALSE))
    )
  }
}



R/performance_mmm.R
# Model performance -------------------------------------------------------

#' Calculate the Area under the Receiver Operating Characteristics Curve
#'
#' Use the timeROC package to calculate the ROC-AUC based on predictions
#' from the mmm_predictions function.
#'
#' @param predictions predictions data from the mmm_predictions function.
#' @param prediction_landmark a numeric vector with the landmark times to select predictions.
#' @param prediction_horizon the prediction horizon *at* which to calculate the ROC-AUC.
#' @param id a character value for the name of the vector of subject identifiers in the predictions data.
#' @param failure_time a character value for the name of the vector with failure times in the predictions data.
#' @param failure a character value for th name of the vector with failure status.
#'
#' @import timeROC
#' @import tidyverse
#'
#' @details Uses the timeROC package to determine the area under the receiver operating characteristics curve.
#'
#' @export
#'
#' @examples
#' # Get the ROC-AUC for a single landmark
#' \dontrun{
#' get_auc(
#'   predictions=predictions,
#'   prediction_landmark=1,
#'   prediction_horizon=5,
#'   id="id",
#'   failure_time="ftime",
#'   failure="failure")
#'
#' # Get the ROC-AUC for a series of landmarks from a list of predictions
#' lm <- 1:5
#' lapply(seq_along(lm), function(l) {
#'   get_auc(
#'     predictions=predictions_test[[l]],
#'     prediction_landmark=lm[l],
#'     prediction_horizon=horizon,
#'     id="transnr",
#'     failure_time="stime",
#'     failure="failure"
#'   )
#' })
#' }
get_auc <- function (
  predictions,
  prediction_landmark,
  prediction_horizon,
  id,
  failure_time,
  failure
) {
  data <- predictions %>%  # This assumes that predictions are always stored in a list of data tables
    unnest(data) %>%
    group_by(get(id)) %>%
    dplyr::select(all_of(c(failure_time, failure))) %>%
    summarize(across(everything(),first)) %>%
    ungroup()
  predict <- predictions %>%
    unnest(prediction) %>%
    group_by(get(id)) %>%
    dplyr::select(-data) %>%
    dplyr::filter(landmark == prediction_landmark,
                  horizon == prediction_horizon) %>%
    ungroup()
  time <- data %>% dplyr::select(all_of(failure_time)) %>% unlist()
  delta <- data %>% dplyr::select(all_of(failure)) %>% unlist()
  marker <- predict$post_fail
  assertthat::assert_that(
    all(
      length(time)==length(delta),
      length(time)==length(marker)
    ),
    msg = paste("The length of the vectors", failure_time, failure,
                "and predictions do not match.")
  )
  roc <- timeROC(T = time,
                 delta = delta,
                 marker = marker,
                 cause = 1,
                 weighting = "marginal",
                 times = prediction_horizon-0.001, # substract a small amount of time to get result
                 iid = TRUE)
  roc <- cbind(roc$Stats, roc$AUC, roc$inference$vect_sd_1, roc$CumulativeIncidence)
  colnames(roc) <- c("failures", "survivors", "censored", "auc", "auc_se", "cuminc")
  roc <- as.data.frame(roc, row.names = FALSE) %>% mutate(landmark = prediction_landmark)
  roc %>%
    dplyr::filter(!is.na(auc)) %>%
    mutate(horizon = prediction_horizon) %>%
    relocate(landmark, horizon, everything())
}

#' Plot predictions from the multivariate mixed model
#'
#' For a given landmark time, plot the probability of failure conditional on survival
#' and longitudinal biomarker evolutions.
#'
#' @param predictions the predictions generated from mmm_predictions()
#' @param outcomes a character vector with the names of the longitudinal outcomes
#' @param id a character value with the name of the subject identifier
#' @param subject a character value with id for subject of interest
#' @param time a scalar value for the time at which the prediction is plotted
#'
#' @import ggplot2
#' @import gridExtra
#' @import tidyverse
#' @import scales
#' @import tidyr
#'
#' @return plots for survival conditional on the longitudinal biomarker trajectory.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' plot_predictions(
#'   predictions=predictions,
#'   outcomes=outcomes,
#'   id="id",
#'   subject="0001",
#'   time="time"
#' )
#' }
plot_predictions <- function (predictions,
                              outcomes,
                              id,
                              subject,
                              time) {
    # FIX: Return marker data to the orginal scale
  # check input
  if((length(subject) == 1) == FALSE) {
    stop("Set `subject` to select a single subject")
  }
  # detect binary outcomes
  is_binary <- sapply(outcomes, function(x, ...) {
    all(
      predictions %>%
        dplyr::select(all_of(id), data) %>%
        tidyr::unnest(data) %>%
        distinct(get(x)) %>%
        unlist() %>%
        sort() == c(0, 1)
    )
  })
  markers <- predictions %>%
    dplyr::select(all_of(id), data) %>%
    dplyr::filter(all_of(id) == subject) %>%
    tidyr::unnest(data)
  # Select prediction data
  pred <- predictions %>%
    dplyr::select(all_of(id), prediction) %>%
    dplyr::filter(all_of(id) == subject) %>%
    tidyr::unnest(prediction)
  # }
  # set-up time axis
  if (nrow(markers) == 0) {
    max_x <- 0
  } else {
    max_x <- markers %>%
      dplyr::select(all_of(time)) %>%
      max()
  }
  # Pivot the data for plotting
  markers_trajectory <- markers %>%
    tidyr::pivot_longer(cols = outcomes, names_to ="outcome")
  markers_trajectory <- markers_trajectory %>%
    dplyr::mutate(is_binary = is_binary[markers_trajectory$outcome])
  # IF binary == TRUE
  binary_plot <- ggplot(
    data = markers_trajectory %>%
      dplyr::filter(
        is_binary == TRUE
        ) %>% as.data.frame(),
    aes(x=time, y=value, col=outcome)) +
    geom_point() +
    geom_line(alpha=0.8) +
    scale_x_continuous(name="",
                       limits=c(0, ceiling(max_x)),
                       breaks=scales::breaks_width(0.5)) +
    scale_y_continuous(
      limits=c(0,1),
      name = "longitudinal outcome") +
    theme_bw() +
    theme(legend.position = "none") +
    facet_wrap(vars(outcome))
  # If binary == FALSE
  continuous_plot <- ggplot(
    data=markers_trajectory %>%
      filter(
        is_binary == FALSE
      ) %>% as.data.frame(),
    aes(x=time, y=value, col=outcome)) +
    geom_point() +
    geom_line(alpha = 0.8) +
    scale_x_continuous(name = "",
                       limits = c(0, ceiling(max_x)),
                       breaks = scales::breaks_width(0.5)
    ) +
    scale_y_continuous(name = "longitudinal outcome") +
    theme_bw() +
    theme(legend.position = "none") +
    facet_wrap(vars(outcome),
               scales="free")
  # Plot the predictions
  surv_plot <- pred %>%
    dplyr::filter(landmark==min(landmark)) %>%
    ggplot(aes(x=horizon,y=post_fail)) +
    geom_line(alpha = 0.8)  +
    scale_x_continuous(name = "",
                       breaks = scales::breaks_width(1)) +
    scale_y_continuous(name = "Probability of graft failure",
                       limits = c(0, 1),
                       breaks = scales::breaks_width(0.2),
                       position = "right") +
    theme_bw()
  # Combine plots and return
  gridExtra::grid.arrange(
    grobs=list(binary_plot, continuous_plot, surv_plot),
    widths=c(1,1),
    layout_matrix=rbind(
      c(1,3),
      c(2,3)
    )
  )
}

#' Plot the calibration for MMM
#'
#' Create a grobs with calibration plot for a given landmark and horizon.
#'
#' @param predictions predictions data from the mmm_predictions function.
#' @param prediction_landmark a numeric vector with the landmark times to select predictions
#' @param prediction_horizon the prediction horizon *at* which to calculate the ROC-AUC.
#' @param id a character value for the name of the vector of subject identifiers in the predictions data
#' @param failure_time a character value for the name of the vector with failure times in the predictions data
#' @param failure a character value with the name of the vector of failure indicators.
#' @param predictions the predictions returned by the mmm_predictions function
#'
#' @import tidyverse
#' @import ggplot2
#' @import ggthemes
#' @import gridExtra
#' @import scales
#' @import riskRegression
#'
#' @return a list of grobs
#'
#' @export
#'
#' @examples
#' \dontrun{
#' # Retrieve calibration data for plotting
#' plot_calibration(
#'   predictions=predictions,
#'   prediction_landmark=1,prediction_horizon=horizon,
#'   id="id",
#'   failure_time="stime",
#'   failure="failure"
#' )
#'
#' # Create a set of calibration plots from a list of predictions
#'   plot_calibration <- lapply(which(lm %in% c(1:3)), function(l) {
#'     plot_calibration(
#'     predictions=predictions[[l]],
#'     prediction_landmark=lm[l],
#'     prediction_horizon=horizon,
#'     id="id",
#'     failure_time="ftime",
#'     failure="failure"
#'   )
#' })
#'
#' # Combine the calibration data
#' plot_calibration <- do.call(rbind, plot_calibration_train)
#'
#' # Create the plot with overlayed calibration lines for each landmark
#' plot_calibration_train <- ggplot(
#'     plot_calibration_train,
#'     aes(x=Pred, y=Obs)
#'   ) +
#'   geom_smooth(
#'     aes(color=factor(landmark),
#'     fill=factor(landmark)),
#'     method="loess",
#'     span=0.3,
#'     alpha=0.2
#'  ) +
#'  geom_abline(intercept=0, slope=1, alpha=0.5) +
#'  geom_rug(aes(x=Pred)) +
#'  scale_y_continuous(breaks=seq(0, 1, 0.2)) +
#'  scale_x_continuous(breaks=seq(0, 1, 0.2)) +
#'  coord_cartesian(
#'    xlim=c(0, 1),
#'    ylim=c(0, 1)
#'  ) +
#'  abs(title="",
#'    caption=paste("Calibration plots for kidney graft failure at",
#'      horizon,
#'      "years follow-up in training data"),
#'    y="Observed risks",
#'    x="Predicted risks",
#'    fill="Landmark"
#'  ) +
#'  guides(color="none") +
#'  ggthemes::theme_tufte(ticks=TRUE)
#'  }

plot_calibration <- function(
  predictions,
  prediction_landmark,
  prediction_horizon,
  id,
  failure_time,
  failure
) {
  data <- predictions %>%  # This assumes that predictions are always stored in a list of data tables
    unnest(data) %>%
    group_by(get(id)) %>%
    dplyr::select(all_of(c(failure_time, failure))) %>%
    summarize(across(everything(),first)) %>%
    ungroup()
  predict <- predictions %>%
    unnest(prediction) %>%
    group_by(get(id)) %>%
    dplyr::select(-data) %>%
    dplyr::filter(landmark == prediction_landmark,
                  horizon == prediction_horizon) %>%
    ungroup()
  score <- riskRegression::Score(
    list( "p_fail"=predict$post_fail ),
    formula=Surv(stime, failure) ~ 1,
    data=data,
    conf.inf=FALSE,
    times=prediction_horizon,
    metrics=NULL,
    plots="Calibration")
  plt_calibrate <- riskRegression::plotCalibration(
    score,
    cens.method="local",
    round=TRUE,
    rug=TRUE,
    plot=FALSE)
  plt_calibrate$plotFrames$p_fail %>%
    dplyr::mutate(landmark=prediction_landmark)
}
